File: neetcode all and misc questions/Arrays and Strings/Two Pointers/merge_strings_alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        res = []
        l = 0
        r = 0
        # to alternate between words, first add word1[l] to res, then word2
        # then increment both pointers
        while (l < len(word1) and r < len(word2)):
            res += word1[l]
            res += word2[r]
            l += 1
            r += 1
        if l == len(word1):
            for k in range(r, len(word2)):
                res.append(word2[k])
        else:
            for k in range(l, len(word1)):
                res.append(word1[k])
        return "".join(res)

File: node_js_ds_and_algos_implementations/arrays and strings/two pointers/merge_strings_alternately.js
class Solution {
    /**
     * @param {string} word1
     * @param {string} word2
     * @return {string}
     */
    mergeAlternately(word1, word2) {
        /* two pointers, 
        get char from word1 then word2, iterate both pointers. Then check to see if one of the pointers hasn't reached the end,
        and append the remaining chars
        O(N) Time
        O(1) Space
        */
        let res = [] 
        let i = 0
        let j = 0
        while (i < word1.length && j < word2.length){
            res.push(word1[i])
            res.push(word2[j])
            ++i
            ++j
        }
        if (i < word1.length){
            for (let k = i; k < word1.length; ++k){
                res.push(word1[k])
            }
        }
        else if (j < word2.length){
            for (let k = j; k < word2.length; ++k){
                res.push(word2[k])
            }
        }
        return res.join("")
    }
}
